pkg/lifecycle/service.go

```diff
@@ -176,6 +185,77 @@ func (s *Service) Start(
 	return nil
 }
 
+// Restart stops an existing pipeline and replaces their nodes.
+func (s *Service) Restart(ctx context.Context, rp *runnablePipeline) error {
+	rp, ok := s.runningPipelines.Get(rp.pipeline.ID)
```

Comment:
Also here, `rp` is the runnable pipeline, I don't think we need to fetch it again.

Reply:
Oops, that's a leftover from when I was using a pipeline instance. Thanks for the catch.